lib/OpenQA/WebAPI/Plugin/ObsRsync/Controller.pm

```diff
@@ -0,0 +1,119 @@
+    return if $self->_check_and_render_error($folder);
+
+    my $full        = $self->_home;
+    my $obs_project = $folder;
+    my $files       = Mojo::File->new($full, $obs_project, '.run_last')->list({dir => 1})->map('basename');
+
+    $self->render(
+        'ObsRsync_folder',
+        obs_project     => $obs_project,
+        lst_files       => $files->grep(qr/files_.*\.lst/)->to_array,
+        read_files_sh   => $files->grep(qr/read_files\.sh/)->join(),
+        rsync_commands  => $files->grep(qr/rsync_.*\.cmd/)->to_array,
+        rsync_sh        => $files->grep(qr/print_rsync.*\.sh/)->to_array,
+        openqa_commands => $files->grep(qr/openqa\.cmd/)->join(),
+        openqa_sh       => $files->grep(qr/print_openqa\.sh/)->join());
+}
+
+sub logs {
+    my $self   = shift;
+    my $folder = $self->param('folder');
+    return if $self->_check_and_render_error($folder);
+
+    my $full = Mojo::File->new($self->_home, $folder);
+    my $files
+      = $full->list({dir => 1, hidden => 1})->map('basename')->grep(qr/\.run_.*/)->sort(sub { $b cmp $a })->to_array;
+    $self->render('ObsRsync_logs', folder => $folder, full => $full->to_string, subfolders => $files);
+}
+
+sub logfiles {
+    my $self      = shift;
+    my $folder    = $self->param('folder');
+    my $subfolder = $self->param('subfolder');
+    return if $self->_check_and_render_error($folder, $subfolder);
+
+    my $full  = Mojo::File->new($self->_home, $folder, $subfolder);
+    my $files = $full->list({dir => 1})->map('basename')->sort->to_array;
+    $self->render(
+        'ObsRsync_logfiles',
+        folder    => $folder,
+        subfolder => $subfolder,
+        full      => $full->to_string,
+        files     => $files
+    );
+}
+
+sub download_file {
+    my $self      = shift;
+    my $folder    = $self->param('folder');
+    my $subfolder = $self->param('subfolder');
+    my $filename  = $self->param('filename');
+    return if $self->_check_and_render_error($folder, $subfolder, $filename);
+
+    my $full = Mojo::File->new($self->_home, $folder, $subfolder);
+    $self->reply->file($full->child($filename));
+}
+
+sub _check_and_render_error {
+    my $self = $_[0];
+    my ($code, $message) = _check_error(@_);
+    $self->render(json => {error => $message}, status => $code) if $code;
+    return $code;
+}
+
+sub _check_error {
+    my $self      = shift;
+    my $project   = shift;
+    my $subfolder = shift;
+    my $filename  = shift;
+    my $home      = $self->_home;
+    return (405, "Home directory is not set") unless $home;
+    return (405, "Home directory not found")  unless -d $home;
+    return (400, "Project has invalid characters")   if $project   && $project =~ m!/!;
+    return (400, "Subfolder has invalid characters") if $subfolder && $subfolder =~ m!/!;
+    return (400, "Filename has invalid characters")  if $filename  && $filename =~ m!/!;
+
+    return (404, "Invalid Project") unless !$project || -d Mojo::File->new($home, $project);
```

Comment:
I personally don't like the use of `unless` when a more complicated expression follows.